Implement generic changelog functionality

kainhofer opened the pull request:
Implement generic changelog functionality: Hook in the log writing directly in the TableAccess base class for all DB records

This is an update to PR #1723, now based on current master (v5). Feature-wise I'm finished. This patch is in a shape for a real review and possible merge.

### High-level Approach explained
By default, whenever a DB record is saved, all modified columns are logged. There is an exclusion list for tables that don't need logging (changelog, autologin, sesssion, etc.), as well as a table-specific exlusion list for column (by default all {usr_id,timestamp}{create,change} columns are ignored, but other tables like _users have some more).

Storing all changes in the adm_log_changes table is implemented. For most use cases (users, user_data, members, profile_fields, events, announcements, roles) the logged data is also properly polished (since there is no generic translation from DB colum to human-readable label, I needed to duplicate a lot of labels from the UI pages). For other DB tables, logging works, but not all DB columns are nicely translated.

Also extended the profile history page to a full changelog page (moved to adm_program/modules/changelog.php) for all different record types, properly linking to most records. Using filters, that page can also show only changes to one (or more) DB tables and records (e.g. changes for one user or one Role or one event).

Also added a change history button to all pages where objects can be modified. The history button will linke to a changelog page that diesplays only the respective DB table or even the particular object identified by UUID).

In the settings, the admin can select whether a changelog should be kept at all, and which tables should be covered.

### Technical details

Since many files are affected, I should probably give a little overview of the relevant code parts:

- The main logging functionality happens in the TableAccess class in the save() and delete() methods, which call helper methods for creation, deletion and modification. They create a TableLogChanges record for the database.
  - Child classes can modify how fields are stored (e.g. the new TableUserData does not lot each of its columns as a separate change, but as one change for the underlying user).
- The new adm_log_changes table holds the ids of the affected records (and potentially related objects, like the group for the members table), as well as human-readable strings from the time of the change. Reasons are that (1) for things like membeships, the underlying Db record might be deleted meanwhile and (2) to me a changelog should be immutable and show the status at the time of the change.
- Most code changes to the modules are just inserting a "change history" button
- The changes to ChangeNotification.php and TableMembers are simply removing the previous logging functionality (but still leave the email-notifications).
- The profile_field_history.php is extended to show changes to all kinds of db tables (filters are passed in as GET variables table, uuid, id and related_id).
- Access to the changelog is governed by the same rules as view/edit permissions: 
  - Admin is always allowed to view changelog
  - Users / Role members with edit access to a particular module can view the module table's changelog (unless changelog is restricted to admin in the preferences)
  - User profile data changes (tables adm_users, adm_user_data, adm_user_relations, adm_members) can be viewed by anyone with editUser permissions. If the changelog for a particular user is requested, individual access (mainly based on role permission) of the current user to the target used is also considered. In the changelog, each change record ist checked separately and hidden if not permitted (a warning message is shown in that case).

### What is implemented

- [x] DB schema creation, DB update from old users_log
- [x] logging each change to the new log table
- [x] per-table preference flag to enable/disable loggin
- [x] Changlog view (based on general, global or individual edit permissions, checked for each individual record), either for all changes, only changes of one or more tables, or one particular object identified by UUID; Links are inserted to jump to the objects or linked objects or DB columns. Also, all IDs are converted to human-readable labels (with links whenever possible).
- [X] Change History button in all overview and edit pages, linking to the log of the current object or object types


### What is still missing

- [ ] Write Documentation how to implement/customize proper logging for core and third-party modules
- [ ] Multi-organization setup
- [ ] Access permissions taking into account Organization ID
- [ ] "Delete History" button in the changelog page (nice-to-have)
- [ ] On the Changelog page, provide a table selection filter, potentially add more filters, e.g. by name or old/new value

### Open/Known Bugs (in my changelog code or in admidio in general)

- [ ] Folder rights (table role_rights, internal_name) are not tanslated / user-readable
- [ ] Deleting a folder also logs removal of each role permission
- [ ] Moving a folder removes and re-adds all role permissions
- [ ] Category administration: Type is not used / stored in changelog (EVT, ANN, …) -> Changelog shows all categories, even when not relevant (e.g. calendars are shown for role categories)
- [ ] 	A lot of Code Duplication in ChangelogService: Translating tables, column names and values to human-readable labels -> no generic function, translations are scattered all over the codebase, whenever a form control is created...
  - [ ] DB column -> readable string conversion from individual record editing pages 
  - [ ] Hardcode the html links to the different kinds of tables
  - [ ] Hardcoded code to generate Entity-derived object for a given DB tble (e.g. new User for 'adm_users' table)
  - [ ] Function formatValue contains a lot of duplicate code from user fields -> Write one generic function that formats values, including links to various forms of objects?
  - [ ] $userFieldText to convert user field types to readable strings

### Tables / Settings that are not yet logged, since they access the DB directly (not via Entity):

- [ ] 	Foto uploads / deletions, => access the filesystem directly, no information stored in DB at all
- [ ] 	Several direct code parts directly accessing the DB (globally search for: "'UPDATE ", "'INSERT ", "'DELETE ")
- [ ] 	RoleDependency  (table role_dependencies) and Preferences classes do not use TableAccess / Entity -> no log written!
- [ ] 	Prefs -> Modules -> Events -> Assign room Checkbox is not logged
- [ ] Events -> setting "I attend" button to "tentative" is saved, but not logged

Fixes Feature requests / Issues #1533, #1069, #326

Fasse commented:
Want to test the functionality:
- empty database
- run demo_data
- run update script

got the following error during update
> VERSION: 5.0.0
> STEP: 600
> 
> S Q L - E R R O R
> 
> SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-changelog_table_members' for key 'adm_preferences.adm_idx_prf_org_id_name'
> 
> CODE: 23000
> B A C K T R A C E
> 
> FILE: src/Infrastructure/Database.php
> LINE: 806
> CALL: Admidio\Infrastructure\Database->showError()
> 
> FILE: src/Components/Entity/ComponentUpdate.php
> LINE: 146
> CALL: Admidio\Infrastructure\Database->queryPrepared()
> 
> FILE: src/Components/Entity/ComponentUpdate.php
> LINE: 200
> CALL: Admidio\Components\Entity\ComponentUpdate->executeUpdateSql()
> 
> FILE: src/Components/Entity/ComponentUpdate.php
> LINE: 274
> CALL: Admidio\Components\Entity\ComponentUpdate->executeStep()
> 
> FILE: src/InstallationUpdate/Service/Update.php
> LINE: 97
> CALL: Admidio\Components\Entity\ComponentUpdate->update()
> 
> FILE: install/update.php
> LINE: 293
> CALL: Admidio\InstallationUpdate\Service\Update->doAdmidioUpdate()

kainhofer commented:
> * empty database
> * run demo_data
> * run update script
> 
> got the following error during update

Hmm, I can't reproduce this here. 
Anyway, I realized I had mistakenly modified the demo data -> reverted. Also, I changed the INSERT INTO in the update script into an REPLACE INTO, so these duplicate key restraint violations should not happen any more.

When you got that error, after completely emptying the DB, did you first go to admidios main page (so that the initial setup screen is shown) or directly to demo_data/? I'm not sure when/if just trying to load the main page will already set up the preferences table. The behavior might have changed significantly due to the revert of the disableExceptions() in the ChangeNotification script.

Fasse commented:
1. The problem is still there. I took a deeper look into the script. The problem is step 590 where you rename a preference to changelog_enable_module. Since this is a new preference set in preferences.php the update script will automatically add this preference to the database. After that the steps are executed and then the index error occured. You can remove 590 and copy the value from the old preference to the new preference and then delete the old preference. Have a look at steps 450, 460 and 470
2. There are problems with PostgreSQL in your statements. Step 570 and step 600 should not have the quotes at the insert column list. That is not valid SQL.
3. REPLACE is not a valid PostgreSQL method. The preference are still there (as of point 1) so you need to update some values. Please use therefore the syntax of steps 450 and 460 and repeat them for each preference.

kainhofer commented:
Thanks for the hints about the preferences and the db update. I have also fixed the pgsql issues.

I realized, I had not enabled debug mode, so I didn't get the warnings you were describing. I've now extended my docker-compose setup to include pgsql and phppgadmin containers, too, so I can test with postgres as well.

I have fixed all issues that you pointed out or that I stumbled across.

Fasse commented on themes/simple/css/admidio.css:
```diff
@@ -609,3 +609,18 @@ tr.admidio-event-highlight {
+/* Changelog configuration  */
+/***********************************/
+
+div.changelog-tableselect-header {
```
Instead of setting this explicitly in the global css file you can set the class `fw-bold` in the template at that place.

Fasse commented on themes/simple/css/admidio.css:
```diff
@@ -609,3 +609,18 @@ tr.admidio-event-highlight {
+    font-weight: bold;
+}
+ul.changelog-tableselect-list {
+    list-style-type: none;
```
Instead of setting this explicitly in the global css file you can set the class `list-unstyled` in the template at that place.

kainhofer replied:
Hmm, list-unstyled not only hides the bullet, but it also removes the indent and it adds vertical padding... 

![image](https://github.com/user-attachments/assets/5f9053b9-f6f5-4ab1-8cfd-a70dfa76f3a1)
![image](https://github.com/user-attachments/assets/3612965c-e02b-4821-bdb9-31aa5ee401f8)

Fasse replied:
Ok, you don't like that style? We can leave it the way you implement it.

Fasse commented on themes/simple/css/admidio.css:
```diff
@@ -609,3 +609,18 @@ tr.admidio-event-highlight {
+ul.changelog-tableselect-list {
+    list-style-type: none;
+}
+ul.changelog-tableselect-list li label.changelog-tableselect-label {
```
Instead of setting this explicitly in the global css file you can set the class `fw-normal` in the template at that place.

kainhofer replied:
Thanks, done.

Fasse reviewed (changes requested):
I have now tested the functionality and looked at the code. Your implementation is really great! Functionally, everything worked and was displayed as I would expect and there were no ambiguities in the operation. Really well done!

The code is also good. Of course I have a few small comments here and there, but these are more organizational things :)

kainhofer commented on adm_program/languages/en.xml:
```diff
@@ -343,6 +358,9 @@
   <string name="SYS_CHANGE_DATE">Change date</string>
   <string name="SYS_CHANGE_HISTORY">Change history</string>
   <string name="SYS_CHANGE_HISTORY_OF">Change history of #VAR1#</string>
+  <string name="SYS_CHANGE_HISTORY_GENERIC">Change history: #VAR1#</string>
```
Actually, what I learnt back in the days when I worked on KDE was that "static (i.e. hardcoded, but not translated) text" that assumes a given word/phrase/separator order is usually not sufficient for some cultures / languages. It's usually better to give the translators a chance to translate or even transform the full phrase rather than individual parts.

Constructing the headline as SYS_CHANGE_HISTORY + ": " + Objectlist will fail if some language does not use the colon as a separator, or when it makes more sense to phrase it as "Objectlist - Change history".

Speaking of, the string before "Change history of #VAR1#" is a particular example that will not work in many non-European languages. In English, there is no difference in the name after an "of", even in German it works with names, since the article is left out. But other languages (e.g. slavic languages) indicate the case in the word's/name's ending rather than a (possible, but for names usually left out article). So for those languages, the value of #VAR1# would need to be adjusted for form a grammatically correct headline.

Fasse replied:
Ok

kainhofer commented on adm_program/languages/en.xml:
```diff
@@ -115,6 +116,18 @@
   <string name="INS_WIFE">Wife</string>
+  <!--Phrases only in Changelog administration-->
+  <!--Phrases only in Changelog administration-->
+  <string name="LOG_ALLOW_DELETION">Allow deletion of the change history</string>
```
Will do.

kainhofer commented on src/Changelog/Service/ChangelogService.php:
```diff
@@ -0,0 +1,846 @@
+        
+        // Changelog disabled globally
+        if ($gSettingsManager->getInt('changelog_enable_module') == 0) {
+            return;
```
Actually, in this case we want the code to do nothing. The displayHistoryButton is always called in inside that function I check whether the corresponding table is logged at all, whether the use has access rights to the table or object, etc.

This allows centralizing the logic when to display the changelog button to this very function instead of scattered all over the various pages. Maybe the function can include a hint that it will not ignorantly insert the button no matter what, but do it's own checks. Would a name like "maybeDisplayHistoryButton" or "displayHistoryButtonIfSuitable" be better?

Fasse replied:
ok

kainhofer commented on src/Changelog/Service/ChangelogService.php:
```diff
@@ -0,0 +1,846 @@
+     * @param array $params
+     * @return void
+     */
+    public static function displayHistoryButton(HTMLPage $page, string $area, string|array $table, bool $condition = true, array $params = array()) : void {
```
I also thought about this. On the other hand, HTMLPage is a generic base class, while displayHistory Button is a very specific use case that depends on the list of DB tables, the current user (or the user that is modified).

Pulling in all this baggage into a generic HTML page base class somehow felt wrong, so I use a dedicated static function in the ChangeLog code parts instead to keep log-related code cleaner together and no pollute a base class. But if you like moving that function to HtmlPage, I can easily do that, too. (Replace all in VS Code is pretty simple...)

Fasse replied:
I had the same thoughts and I'm not sure which is the best solution. Maybe we leave it the way you implement it.

kainhofer commented on install/db_scripts/update_5_0.xml:
```diff
@@ -61,5 +61,76 @@ gender-trans|SYS_DIVERSE' WHERE usf_name_intern = 'GENDER' AND usf_value_list LI
+       <!-- rename preference 'profile_log_edit_fields' to 'changelog_enable_module' -->
+    <step id="590" database="mysql">UPDATE %PREFIX%_preferences pr1 INNER JOIN %PREFIX%_preferences pr2 ON pr2.prf_name = 'profile_log_edit_fields' SET pr1.prf_value = pr2.prf_value WHERE pr1.prf_name = 'changelog_enable_module'</step>
+    <step id="600" database="pgsql">UPDATE %PREFIX%_preferences pr1 SET prf_value = pr2.prf_value FROM %PREFIX%_preferences pr2 WHERE pr2.prf_name = 'profile_log_edit_fields' AND pr1.prf_name = 'changelog_enable_module'</step>
+    <step id="610">DELETE FROM %PREFIX%_preferences WHERE prf_name = 'profile_log_edit_fields'</step>
```
Actually, I'm not so sure we want this constraint in the log table. If the User who did the original change is deleted (e.g. because a secretary or a volunteer has left), I don't want to set the log_usr_id_create to 0, effectively removing the information who made the change. Of course, the name is no longer there and recoverable, but at least it is possible to find all changes made by that user. That's quite important if you need to clean up your member base after a sub-optimal volunteer hast left.

Especially if you have a large fluctuation of helpers or administrative personnell, it's very usefull (and for audit purposes even neccessary) to distinguish changes made by different persons that have left.

Fasse replied:
Ok, that could be an argument. We can start without that constraint.

Fasse commented:
@kainhofer I have merged your changes. Do you want to write a documentation in the wiki about the use of the history / log feature?

kainhofer commented:
Yes, documenting how the changelog works / how it can be tweaked for other modules is on my list.

As written in my new PR with further improvements, I have some open questions:

- Are you okay with the massive duplication of DB column and table name translations in ChangelogService.php? Most user-readable labels are "hardcoded" in the preferences or module forms, so I couldn't see a better way to get the labels than duplicating them in the changelog page.
- The current implementation is NOT designed for a multi-organization setup. Creating/deleting an organization is logged, but in the changelog view and the DB table, there is no restriction on the current org. I have no experience with multiple organizations, so maybe you have some ideas how to best handle this situation.
- Performance: The changelog page currently loads and formats all changelog entries in the requested time frame. For large installations, this might overwhelm the browser and/or network. Maybe the changelog should be switched to dynamic page loading (like the contact list)? What do you think? Do you have any experience with performance?
